File: src/mods/stage_edits.cpp
#include "stage_edits.h"
#include "systems/pref.h"
#include "utils/patch.h"

namespace stage_edits {

enum class ActiveMode {
    None = 0,
    Golden = 1,
    Dark = 2,
    Reverse = 3,
};

static ActiveMode s_current_mode = ActiveMode::None;
static u32 s_rev_goal_idx = 0;
static bool s_new_goal = false;

static patch::Tramp<decltype(&mkb::smd_game_ready_init)> s_ready_init_tramp;
static patch::Tramp<decltype(&mkb::load_stagedef)> s_load_stagedef_tramp;

void select_new_goal() { s_new_goal = true; }

static void undo_mode(ActiveMode mode) {
    switch (mode) {
        case ActiveMode::None: {
            break;
        }
        case ActiveMode::Golden: {
            // disable goals somehow
            for (u32 i = 0; i < mkb::stagedef->goal_count; i++) {
                mkb::stagedef->goal_list[i].position.y += 10000;
            }
            break;
        }
        case ActiveMode::Dark: {
            break;
        }
        case ActiveMode::Reverse: {
            if (mkb::stagedef->goal_count < 1) return;
            float x = mkb::stagedef->start->position.x;
            float y = mkb::stagedef->start->position.y;
            float z = mkb::stagedef->start->position.z;
            s16 sx = mkb::stagedef->start->rotation.x;
            s16 sy = mkb::stagedef->start->rotation.y;
            s16 sz = mkb::stagedef->start->rotation.z;

            mkb::stagedef->start->position.x = mkb::stagedef->goal_list[s_rev_goal_idx].position.x;
            mkb::stagedef->start->position.y =
                mkb::stagedef->goal_list[s_rev_goal_idx].position.y + 0.5;
            mkb::stagedef->start->position.z = mkb::stagedef->goal_list[s_rev_goal_idx].position.z;
            mkb::stagedef->start->rotation.x = mkb::stagedef->goal_list[s_rev_goal_idx].rotation.x;
            mkb::stagedef->start->rotation.y =
                mkb::stagedef->goal_list[s_rev_goal_idx].rotation.y + 32766;
            mkb::stagedef->start->rotation.z = mkb::stagedef->goal_list[s_rev_goal_idx].rotation.z;

            mkb::stagedef->goal_list[s_rev_goal_idx].position.x = x;
            mkb::stagedef->goal_list[s_rev_goal_idx].position.y = y - 0.5;
            mkb::stagedef->goal_list[s_rev_goal_idx].position.z = z;
            mkb::stagedef->goal_list[s_rev_goal_idx].rotation.x = sx;
            mkb::stagedef->goal_list[s_rev_goal_idx].rotation.y = sy - 32766;
            mkb::stagedef->goal_list[s_rev_goal_idx].rotation.z = sz;
            break;
        }
    }
}

static void set_mode(ActiveMode mode) {
    switch (mode) {
        case ActiveMode::None: {
            break;
        }
        case ActiveMode::Golden: {
            // disable goals somehow
            for (u32 i = 0; i < mkb::stagedef->goal_count; i++) {
                mkb::stagedef->goal_list[i].position.y -= 10000;
            }
            break;
        }
        case ActiveMode::Dark: {
            break;
        }
        case ActiveMode::Reverse: {
            if (mkb::stagedef->goal_count < 1) return;
            s_rev_goal_idx %= mkb::stagedef->goal_count;
            // switch goal and start
            float x = mkb::stagedef->start->position.x;
            float y = mkb::stagedef->start->position.y;
            float z = mkb::stagedef->start->position.z;
            s16 sx = mkb::stagedef->start->rotation.x;
            s16 sy = mkb::stagedef->start->rotation.y;
            s16 sz = mkb::stagedef->start->rotation.z;

            mkb::stagedef->start->position.x = mkb::stagedef->goal_list[s_rev_goal_idx].position.x;
            mkb::stagedef->start->position.y =
                mkb::stagedef->goal_list[s_rev_goal_idx].position.y + 0.5;
            mkb::stagedef->start->position.z = mkb::stagedef->goal_list[s_rev_goal_idx].position.z;
            mkb::stagedef->start->rotation.x = mkb::stagedef->goal_list[s_rev_goal_idx].rotation.x;
            mkb::stagedef->start->rotation.y =
                mkb::stagedef->goal_list[s_rev_goal_idx].rotation.y + 32766;
            mkb::stagedef->start->rotation.z = mkb::stagedef->goal_list[s_rev_goal_idx].rotation.z;

            mkb::stagedef->goal_list[s_rev_goal_idx].position.x = x;
            mkb::stagedef->goal_list[s_rev_goal_idx].position.y = y - 0.5;
            mkb::stagedef->goal_list[s_rev_goal_idx].position.z = z;
            mkb::stagedef->goal_list[s_rev_goal_idx].rotation.x = sx;
            mkb::stagedef->goal_list[s_rev_goal_idx].rotation.y = sy - 32766;
            mkb::stagedef->goal_list[s_rev_goal_idx].rotation.z = sz;
            break;
        }
    }
}

void main_game_init() {
    patch::hook_function(s_ready_init_tramp, mkb::smd_game_ready_init, []() {
        ActiveMode next_mode = ActiveMode(pref::get(pref::U8Pref::StageEditVariant));
        if (s_current_mode != next_mode) {
            undo_mode(s_current_mode);
            s_current_mode = ActiveMode(pref::get(pref::U8Pref::StageEditVariant));
            set_mode(s_current_mode);
        } else if (s_current_mode == ActiveMode::Reverse && s_new_goal) {
            undo_mode(ActiveMode::Reverse);
            s_rev_goal_idx++;
            set_mode(ActiveMode::Reverse);
        }
        s_new_goal = false;
        s_ready_init_tramp.dest();
    });
}

void init() {
    patch::hook_function(s_load_stagedef_tramp, mkb::load_stagedef, [](u32 stage_id) {
        s_load_stagedef_tramp.dest(stage_id);
        s_current_mode = ActiveMode(pref::get(pref::U8Pref::StageEditVariant));
        set_mode(s_current_mode);
    });
}

void tick() {
    switch (s_current_mode) {
        case ActiveMode::None: {
            break;
        }
        case ActiveMode::Golden: {
            if (mkb::mode_info.bananas_remaining == 0) {
                mkb::mode_info.g_ball_mode |= 0x228;
            }
            break;
        }
        case ActiveMode::Dark: {
            if (mkb::stagedef != nullptr &&
                mkb::mode_info.bananas_remaining != mkb::stagedef->banana_count) {
                mkb::mode_info.g_ball_mode |= mkb::BALLMODE_FALLEN_OUT;
            }
            break;
        }
        case ActiveMode::Reverse: {
            break;
        }
    }
}

}  // namespace stage_edits

File: src/mods/stage_edits.rs
#![cfg(feature = "mkb2")]

use num_enum::TryFromPrimitive;

use crate::mkb2::mkb2;
use crate::{app::AppContext, hook, systems::pref::U8Pref};

hook!(LoadStagedefHook, stage_id: u32 => (), mkb2::load_stagedef, |stage_id, cx| {
    cx.stage_edits.borrow_mut().on_load_stagedef(stage_id, cx);
});

#[derive(Clone, Copy, PartialEq, Eq, TryFromPrimitive, Default)]
#[repr(u8)]
pub enum ActiveMode {
    #[default]
    None = 0,
    Golden = 1,
    Dark = 2,
    Reverse = 3,
}

#[derive(Default)]
pub struct StageEdits {
    current_mode: ActiveMode,
    rev_goal_idx: u32,
    new_goal: bool,
    load_stagedef_hook: LoadStagedefHook,
}

impl StageEdits {
    pub fn on_main_loop_load(&mut self, _cx: &AppContext) {
        self.load_stagedef_hook.hook();
    }

    pub fn select_new_goal(&mut self) {
        self.new_goal = true;
    }

    unsafe fn undo_mode(&self, mode: ActiveMode) {
        match mode {
            ActiveMode::None => {}
            ActiveMode::Golden => {
                // disable goals somehow
                for i in 0..((*mkb2::stagedef).goal_count as usize) {
                    (*(*mkb2::stagedef).goal_list.add(i)).position.y += 10000.0;
                }
            }
            ActiveMode::Dark => {}
            ActiveMode::Reverse => {
                if (*mkb2::stagedef).goal_count < 1 {
                    return;
                }

                let start = (*mkb2::stagedef).start;
                let goal = (*mkb2::stagedef).goal_list.add(self.rev_goal_idx as usize);

                let x = (*start).position.x;
                let y = (*start).position.y;
                let z = (*start).position.z;
                let sx = (*start).rotation.x;
                let sy = (*start).rotation.y;
                let sz = (*start).rotation.z;

                (*start).position.x = (*goal).position.x;
                (*start).position.y = (*goal).position.y + 0.5;
                (*start).position.z = (*goal).position.z;
                (*start).rotation.x = (*goal).rotation.x;
                (*start).rotation.y = (*goal).rotation.y + 32766;
                (*start).rotation.z = (*goal).rotation.z;

                (*goal).position.x = x;
                (*goal).position.y = y - 0.5;
                (*goal).position.z = z;
                (*goal).rotation.x = sx;
                (*goal).rotation.y = sy - 32766;
                (*goal).rotation.z = sz;
            }
        }
    }

    unsafe fn set_mode(&mut self, mode: ActiveMode) {
        match mode {
            ActiveMode::None => {}
            ActiveMode::Golden => {
                // disable goals somehow
                for i in 0..((*mkb2::stagedef).goal_count as usize) {
                    (*(*mkb2::stagedef).goal_list.add(i)).position.y -= 10000.0;
                }
            }
            ActiveMode::Dark => {}
            ActiveMode::Reverse => {
                if (*mkb2::stagedef).goal_count < 1 {
                    return;
                }
                self.rev_goal_idx %= (*mkb2::stagedef).goal_count as u32;

                let start = (*mkb2::stagedef).start;
                let goal = (*mkb2::stagedef).goal_list.add(self.rev_goal_idx as usize);

                let x = (*start).position.x;
                let y = (*start).position.y;
                let z = (*start).position.z;
                let sx = (*start).rotation.x;
                let sy = (*start).rotation.y;
                let sz = (*start).rotation.z;

                (*start).position.x = (*goal).position.x;
                (*start).position.y = (*goal).position.y + 0.5;
                (*start).position.z = (*goal).position.z;
                (*start).rotation.x = (*goal).rotation.x;
                (*start).rotation.y = (*goal).rotation.y + 32766;
                (*start).rotation.z = (*goal).rotation.z;

                (*goal).position.x = x;
                (*goal).position.y = y - 0.5;
                (*goal).position.z = z;
                (*goal).rotation.x = sx;
                (*goal).rotation.y = sy - 32766;
                (*goal).rotation.z = sz;
            }
        }
    }

    fn on_load_stagedef(&mut self, stage_id: u32, cx: &AppContext) {
        self.load_stagedef_hook.call(stage_id);

        let pref = &cx.pref.borrow();
        let next_mode = ActiveMode::try_from(pref.get_u8(U8Pref::StageEditVariant)).unwrap();
        self.current_mode = next_mode;
        unsafe {
            self.set_mode(self.current_mode);
        }
    }

    pub fn on_game_ready_init(&mut self, cx: &AppContext) {
        let pref = &cx.pref.borrow();
        let next_mode = ActiveMode::try_from(pref.get_u8(U8Pref::StageEditVariant)).unwrap();
        if self.current_mode != next_mode {
            unsafe {
                self.undo_mode(self.current_mode);
                self.current_mode = next_mode;
                self.set_mode(self.current_mode);
            }
        } else if self.current_mode == ActiveMode::Reverse && self.new_goal {
            unsafe {
                self.undo_mode(ActiveMode::Reverse);
                self.rev_goal_idx += 1;
                self.set_mode(ActiveMode::Reverse);
            }
        }
        self.new_goal = false;
    }

    pub fn tick(&mut self, _cx: &AppContext) {
        unsafe {
            match self.current_mode {
                ActiveMode::None => {}
                ActiveMode::Golden => {
                    if mkb2::mode_info.bananas_remaining == 0 {
                        mkb2::mode_info.ball_mode |= 0x228;
                    }
                }
                ActiveMode::Dark => {
                    if !mkb2::stagedef.is_null()
                        && mkb2::mode_info.bananas_remaining != (*mkb2::stagedef).banana_count
                    {
                        mkb2::mode_info.ball_mode |= mkb2::BALLMODE_FALLEN_OUT;
                    }
                }
                ActiveMode::Reverse => {}
            }
        }
    }
}
